File: backend/internal/dto/validation_run.go
package dto

import (
	"encoding/json"
	"time"
)

type CollisionEvent struct {
	SegmentIndex    int     `json:"segment_index"`
	FirstTimeMS     float64 `json:"first_time_ms"`
	ZoneID          uint    `json:"zone_id"`
	ZoneName        string  `json:"zone_name"`
	ZoneType        string  `json:"zone_type"`
	AllowedSpeedMMS float64 `json:"allowed_speed_mm_s"`
	ActualSpeedMMS  float64 `json:"actual_speed_mm_s"`
	ClearanceMM     float64 `json:"clearance_mm"`
	Violation       bool    `json:"violation"`
	Evidence        string  `json:"evidence"`
}

type InterlockFinding struct {
	Code      string   `json:"code"`
	Event     string   `json:"event"`
	DependsOn string   `json:"depends_on,omitempty"`
	Path      []string `json:"path,omitempty"`
	Evidence  string   `json:"evidence"`
}

type CreateValidationRunRequest struct {
	MotionProgramID uint `json:"motion_program_id" validate:"required"`
	RetryFailed     bool `json:"retry_failed"`
}

type ReviewValidationRequest struct {
	Note string `json:"note" validate:"required,min=8,max=1000"`
}

type ValidationRunResponse struct {
	ID                uint               `json:"id"`
	MotionProgramID   uint               `json:"motion_program_id"`
	ProgramCode       string             `json:"program_code"`
	ProgramVersion    int                `json:"program_version"`
	ZoneSnapshot      json.RawMessage    `json:"zone_snapshot"`
	ProgramSnapshot   json.RawMessage    `json:"program_snapshot"`
	AlgorithmVersion  string             `json:"algorithm_version"`
	InputHash         string             `json:"input_hash"`
	IdempotencyKey    string             `json:"idempotency_key"`
	Attempt           int                `json:"attempt"`
	RetryOfID         *uint              `json:"retry_of_id"`
	CollisionEvents   []CollisionEvent   `json:"collision_events"`
	InterlockFindings []InterlockFinding `json:"interlock_findings"`
	RiskScore         float64            `json:"risk_score"`
	ValidationStatus  string             `json:"validation_status"`
	Explanation       string             `json:"explanation"`
	RequestedBy       uint               `json:"requested_by"`
	StartedAt         time.Time          `json:"started_at"`
	FinishedAt        *time.Time         `json:"finished_at"`
	ReviewedBy        *uint              `json:"reviewed_by"`
	ReviewedAt        *time.Time         `json:"reviewed_at"`
	ReviewNote        string             `json:"review_note"`
	Reused            bool               `json:"reused"`
}

File: frontend/src/types/validation-run.ts
import { ValidationStatus } from './enums/validation-status';
import { InterlockEvent, TrajectoryPoint } from './motion-program';
import { SafetyZone } from './safety-zone';

export interface CollisionEvent {
  segment_index: number;
  first_time_ms: number;
  zone_id: number;
  zone_name: string;
  zone_type: string;
  allowed_speed_mm_s: number;
  actual_speed_mm_s: number;
  clearance_mm: number;
  violation: boolean;
  evidence: string;
}

export interface InterlockFinding {
  code: string;
  event: string;
  depends_on?: string;
  path?: string[];
  evidence: string;
}

export interface ProgramSnapshot {
  id: number;
  robot_cell_id: number;
  program_code: string;
  version: number;
  trajectory?: TrajectoryPoint[];
  tool_radius_mm: number;
  payload_radius_mm: number;
  interlock_sequence: InterlockEvent[];
  source_checksum: string;
  program_state: string;
}

export type ZoneSnapshot = Pick<SafetyZone, 'id' | 'name' | 'zone_type' | 'polygon_geojson' | 'min_height_mm' | 'max_height_mm' | 'speed_limit_mm_s' | 'access_rule' | 'version'>;

export interface ValidationRun {
  id: number;
  motion_program_id: number;
  program_code: string;
  program_version: number;
  zone_snapshot: ZoneSnapshot[];
  program_snapshot: ProgramSnapshot;
  algorithm_version: string;
  input_hash: string;
  idempotency_key: string;
  attempt: number;
  retry_of_id?: number;
  collision_events: CollisionEvent[];
  interlock_findings: InterlockFinding[];
  risk_score: number;
  validation_status: ValidationStatus;
  explanation: string;
  requested_by: number;
  started_at: string;
  finished_at?: string;
  reviewed_by?: number;
  reviewed_at?: string;
  review_note: string;
  reused: boolean;
}

export interface AuditEvent {
  id: number;
  actor: string;
  role: string;
  action: string;
  resource_type: string;
  resource_id: string;
  request_id: string;
  parameters: Record<string, unknown>;
  before: Record<string, unknown>;
  after: Record<string, unknown>;
  created_at: string;
}
